File: internal/cli/output/csv.go
package output

import (
	"io"
	"os"
)

// nopCloseWriter wraps an io.Writer with a no-op Close method.
type nopCloseWriter struct {
	io.Writer
}

// Close is a no-op for nopCloseWriter.
func (n *nopCloseWriter) Close() error { return nil }

// OpenExportOutput opens an output destination for export data.
// If path is empty or "-", returns a wrapper around os.Stdout whose Close is a no-op.
// Otherwise creates or truncates the file at path.
func OpenExportOutput(path string) (io.WriteCloser, error) {
	if path == "" || path == "-" {
		return &nopCloseWriter{Writer: os.Stdout}, nil
	}
	return os.Create(path)
}

File: demo/web/packages/xyncra-client-cli/src/output/csv.ts
/**
 * CSV/JSON export output helper.
 *
 * Mirrors Go internal/cli/output/csv.go.
 *
 * @module
 */

import { createWriteStream, type WriteStream } from 'node:fs';
import type { Writable } from 'node:stream';

/**
 * A writable output destination that can be closed.
 * When path is empty or "-", writes to stdout (close is no-op).
 */
export interface ExportOutput {
  writer: Writable;
  close(): void;
}

/**
 * Open an output destination for export data.
 * If path is empty or "-", returns stdout. Otherwise creates/truncates the file.
 */
export function openExportOutput(path: string): ExportOutput {
  if (!path || path === '-') {
    return {
      writer: process.stdout,
      close() {
        // no-op for stdout
      },
    };
  }
  const stream: WriteStream = createWriteStream(path);
  return {
    writer: stream,
    close() {
      stream.end();
    },
  };
}
